sql/src/main/java/io/crate/operation/projectors/BatchIteratorBackpressureExecutor.java

```diff
@@ -0,0 +1,198 @@
+            } else {
+                executionFuture.completeExceptionally(t);
+            }
+        };
+    }
+
+    private BiConsumer<R, Throwable> createBulkExecutionCompleteListener(BatchIterator batchIterator) {
+        return (r, t) -> {
+            if (lastBulkScheduledToExecute &&
+                pendingItemsCount.get() == 0 &&
+                computeFinalResult.compareAndSet(false, true)) {
+
+                // all bulks are complete, close iterator and complete executionFuture
+                batchIterator.close();
+                if (t == null) {
+                    executionFuture.complete(null);
+                } else {
+                    executionFuture.completeExceptionally(t);
+                }
+            } else {
+                safeConsumeIterator(batchIterator);
+            }
+        };
+    }
+
+    /**
+     * Consumes the iterator only if there isn't another thread already consuming it.
+     */
+    private void safeConsumeIterator(BatchIterator batchIterator) {
+        if (collectingEnabled.compareAndSet(false, true)) {
+            unsafeConsumeIterator(batchIterator);
+        }
+    }
+
+    /**
+     * Consumes the iterator without guarding against concurrent access to the iterator.
+     *
+     * !! THIS SHOULD ONLY BE CALLED BY THE loadNextBatch COMPLETE LISTENER !!
+     */
+    private void unsafeConsumeIterator(BatchIterator batchIterator) {
+        Row row = RowBridging.toRow(batchIterator.rowData());
+        try {
+            while (true) {
+                if (indexInBulk == bulkSize) {
+                    if (tryExecuteBulk(false) == false) {
+                        collectingEnabled.set(false);
+                        if (isScheduledCollectionRunning.compareAndSet(false, true)) {
+                            scheduleConsumeIterator();
+                        }
+                        return;
+                    }
+                }
+
+                if (batchIterator.moveNext()) {
+                    indexInBulk++;
+                    onRowConsumer.accept(row);
+                    pendingItemsCount.incrementAndGet();
+                } else {
+                    break;
+                }
+            }
+
+            if (batchIterator.allLoaded()) {
+                lastBulkScheduledToExecute = true;
+                executeFunction.apply(true).whenComplete(bulkExecutionCompleteListener);
+            } else {
+                batchIterator.loadNextBatch().whenComplete(loadNextBatchCompleteListener);
+            }
+        } catch (Throwable t) {
+            batchIterator.close();
+            executionFuture.completeExceptionally(t);
+        }
+    }
+
+    private void scheduleConsumeIterator() throws EsRejectedExecutionException {
+        scheduler.schedule(scheduleConsumeIteratorJob,
+            consumeIteratorDelays.next().getMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    private boolean tryExecuteBulk(boolean isLastBatch) {
```

Comment:
isLastBatch is always false - seems a bit odd given that the ShardingUpsertExecutor uses the isLastBatch flag?